### ISSUE TITLE
_nest() call unexpected result

### ISSUE THREAD

**Issue by AndrewJDR** [dvidelabs/flatcc#52]
Another head scratcher for me.

Given this:
```c
#include <stdio.h>
#include "qmsg.h"

#undef ns
#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(QQQ, x)

int main()
{
  size_t nestedBufSize;
  uint8_t nestedBuffer[1024];

  // Build and store the nested buffer
  {
    flatcc_builder_t pBuilder;
    flatcc_builder_init(&pBuilder);
    ns(Payload_start_as_root(&pBuilder));
      ns(Payload_someId_add(&pBuilder, 456));
    ns(Payload_end_as_root(&pBuilder));

    nestedBufSize = flatcc_builder_get_buffer_size(&pBuilder);
    flatcc_builder_copy_buffer(&pBuilder, nestedBuffer, sizeof(nestedBuffer));
    flatcc_builder_clear(&pBuilder);
  }

  {
    printf("nested buffer: %p nested buffer size:%u\n", nestedBuffer, (uint32_t)nestedBufSize);
    ns(Payload_table_t) nestedPayload = ns(Payload_as_root(nestedBuffer));
    uint32_t someId_before_nesting = ns(Payload_someId(nestedPayload));
    printf("someId (before nesting): %u\n", someId_before_nesting); // <--- prints 456 as expected
  }

  {
    flatcc_builder_t builder;
    flatcc_builder_init(&builder);

    ns(Msg_start_as_root(&builder));
      //ns(Msg_payload_start_as_root(&builder)); <-- also tried with payload_start/end_as_root, prints "12" for someId after nesting.
      ns(Msg_payload_nest(&builder, nestedBuffer, nestedBufSize, 0)); // <--- also tried with 4 and 8 for align. Still prints 0
      //ns(Msg_payload_end_as_root(&builder));
    ns(Msg_end_as_root(&builder));

    size_t bufSize;
    uint8_t *bufPtr = (uint8_t*)flatcc_builder_get_direct_buffer(&builder, &bufSize);
    ns(Msg_table_t) msgTable = ns(Msg_as_root(bufPtr));
    ns(Payload_table_t) nestedPayload = ns(Msg_payload_as_root(msgTable));

    uint32_t someId = ns(Payload_someId(nestedPayload));
    printf("someId (after nesting): %u\n", someId); // <--- Why does this print 0 instead of 456?
  }

  return 0;
}
```

With this schema:
```
namespace QQQ;

table Payload
{
  someId:uint;
}

table Msg
{
  payload:[ubyte] (nested_flatbuffer: "Payload");
}

root_type Msg;
```

I'm expecting this line:
```c
printf("someId (after nesting): %u\n", someId);
```
To print: "456". Any idea why it's printing 0 instead? Thank you.

**Comment by AndrewJDR** [comment]
*edit*
Changed Msg_payload_start_as_root(&pBuilder)/Msg_payload_end_as_root(&pBuilder) calls to Payload_start_as_root(&pBuilder)/Payload_end_as_root(&pBuilder).

These were not quite the appropriate calls to make since I'm not making a nested buffer at that point, I'm just making a standalone one that will later be nested. At first I thought this was my problem, but ultimately even after making the change, it made no difference in the results from what I could see.

**Comment by mikkelfj** [comment]
When testing, could you please try call verify on the buffer with a debug linked library(flatccrt_d)
https://github.com/dvidelabs/flatcc#verifying-a-buffer
and the new section I just wrote:
https://github.com/dvidelabs/flatcc#debugging-a-buffer

I added a CMake option to enable traces - haven't really testet this recently, but might help.

**Comment by AndrewJDR** [comment]
Sure. That does indeed produce an error: `Msg buffer is invalid: vtable offset out of range or unaligned`

Updated code:
```
#include <stdio.h>
#include "qmsg.h"

#undef ns
#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(QQQ, x)

int main()
{
  size_t nestedBufSize;
  uint8_t nestedBuffer[1024];

  // Build and store the nested buffer
  {
    flatcc_builder_t pBuilder;
    flatcc_builder_init(&pBuilder);
    ns(Payload_start_as_root(&pBuilder));
      ns(Payload_someId_add(&pBuilder, 456));
    ns(Payload_end_as_root(&pBuilder));

    nestedBufSize = flatcc_builder_get_buffer_size(&pBuilder);
    flatcc_builder_copy_buffer(&pBuilder, nestedBuffer, sizeof(nestedBuffer));
    flatcc_builder_clear(&pBuilder);
  }

  {
    printf("nested buffer: %p nested buffer size:%u\n", nestedBuffer, (uint32_t)nestedBufSize);
    ns(Payload_table_t) nestedPayload = ns(Payload_as_root(nestedBuffer));
    uint32_t someId_before_nesting = ns(Payload_someId(nestedPayload));
    printf("someId (before nesting): %u\n", someId_before_nesting); // <--- prints 456 as expected
  }

  {
    flatcc_builder_t builder;
    flatcc_builder_init(&builder);

    ns(Msg_start_as_root(&builder));
      //ns(Msg_payload_start_as_root(&builder)); <-- also tried with payload_start/end_as_root, prints "12" for someId after nesting.
      ns(Msg_payload_nest(&builder, nestedBuffer, nestedBufSize, 0)); // <--- also tried with 4 and 8 for align.
      //ns(Msg_payload_end_as_root(&builder));
    ns(Msg_end_as_root(&builder));

    size_t bufSize;
    uint8_t *bufPtr = (uint8_t*)flatcc_builder_get_direct_buffer(&builder, &bufSize);
    size_t finalBufferSize = flatcc_builder_get_buffer_size(&builder);
    printf("final buffer size: %u\n", (uint32_t)finalBufferSize);
    ns(Msg_table_t) msgTable = ns(Msg_as_root(bufPtr));
    ns(Payload_table_t) nestedPayload = ns(Msg_payload_as_root(msgTable));

    uint32_t someId = ns(Payload_someId(nestedPayload));
    printf("someId (after nesting): %u\n", someId); // <--- Why doesn't this print 456?

    int32_t ret;
    if ((ret = ns(Msg_verify_as_root(bufPtr, finalBufferSize)))) {
      printf("Msg buffer is invalid: %s\n",
      flatcc_verify_error_string(ret));
    }

  }

  return 0;
}
```

**Comment by AndrewJDR** [comment]
I note that if I use Msg_payload_start_as_root/Msg_payload_end_as_root calls around the _nest() call, I no longer get an error upon verify. However, as noted previously, the incorrect value of "12" is printed for someId when I do this. My understanding of the interface is that you're not supposed to use start_as_root/end_as_root around a _nest() call, so I'm not surprised that this is giving an odd result.

At the moment I'm still not sure why this is happening.

**Comment by AndrewJDR** [comment]
backtrace from gdb with those FLATCC_DEBUG_VERIFY/FLATCC_TRACE_VERIFY flags enabled:
```
verifier assert: vtable offset out of range or unaligned
fbs_crash: flatcc/g/src/runtime/verifier.c:308: verify_table: Assertion `0' failed.

Program received signal SIGABRT, Aborted.
─── Stack ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
[0] from 0x00007ffff7a4bcc9 in raise+57
[1] from 0x00007ffff7a4f0d8 in abort+328
[2] from 0x00007ffff7a44b86
[3] from 0x00007ffff7a44c32 in __assert_fail+66
[4] from 0x0000000000407617 in verify_table+382 at flatcc/g/src/runtime/verifier.c:308
[5] from 0x0000000000408556 in flatcc_verify_table_as_nested_root+318 at flatcc/g/src/runtime/verifier.c:498
[6] from 0x00000000004010af in __QQQ_Msg_table_verifier+51 at qmsg.h:1066
[7] from 0x000000000040789a in verify_table+1025 at flatcc/g/src/runtime/verifier.c:324
[8] from 0x00000000004082a3 in flatcc_verify_table_as_root+137 at flatcc/g/src/runtime/verifier.c:454
[9] from 0x00000000004010ee in QQQ_Msg_verify_as_root+42 at qmsg.h:1072
[10] from 0x00000000004012b0 in main+448 at fbs_unexpected.c:73
```

**Comment by AndrewJDR** [comment]
Some hexdumps:
```
buffer (with inline calls):
  0000  04 00 00 00 e0 ff ff ff  04 00 00 00 14 00 00 00  ................
  0010  04 00 00 00 f6 ff ff ff  c8 01 00 00 00 00 06 00  ................
  0020  08 00 04 00 06 00 08 00  04 00                    ..........
with _nest():
  0000  04 00 00 00 d4 ff ff ff  04 00 00 00 20 00 00 00  ............ ...
  0010  04 00 00 00 12 00 00 00  04 00 00 00 f8 ff ff ff  ................
  0020  c8 01 00 00 06 00 08 00  04 00 00 00 00 00 00 00  ................
  0030  06 00 08 00 04 00                                ......
```

"inline calls" means:
```
  {
    flatcc_builder_t builder;
    flatcc_builder_init(&builder);
    ns(Msg_start_as_root(&builder));
      ns(Msg_payload_start_as_root(&builder));
	ns(Payload_someId_add(&builder, 456));
      ns(Msg_payload_end_as_root(&builder));
    ns(Msg_end_as_root(&builder));

    size_t bufSize;
    uint8_t *bufPtr = (uint8_t*)flatcc_builder_get_direct_buffer(&builder, &bufSize);
    hexdump("buffer (with inline calls)", bufPtr, bufSize, stdout);
  }
```

and _nest() means using the code from the OP above.

**Comment by AndrewJDR** [comment]
I feel like the _nest() version and the inline calls version should produce identical results, but they're fairly different...

**Comment by mikkelfj** [comment]
Added https://github.com/dvidelabs/flatcc/blob/master/doc/binary-format.md in case you or someone else wants to eyeball the dump.

I will have a look at the issue soon.

**Comment by AndrewJDR** [comment]
@mikkelfj Thanks, very useful.

**Comment by mikkelfj** [comment]
> I feel like the _nest() version and the inline calls version should produce identical results, but they're fairly different...

I can't really make sense of the nest version right now, but they are likely different because a top-level buffer will seek to place vtables at the end (clustered vtables) whereas nested buffers built inside a parent buffer has not space to place the vtable - it cannot extend upwards without running into the parent content. So it places the vtable earlier instead. It doesn't matter which direction only that the buffer is selfcontained with the vector it is placed - but details on what is going must wait till tomorrow.

**Comment by AndrewJDR** [comment]
Test and PR that hopefully fixes this over at #53.

**Comment by mikkelfj** [comment]
Thanks!
